contrib/babelfishpg_tsql/sql/sys_views.sql

```diff
@@ -1960,17 +1960,19 @@ GRANT SELECT ON sys.endpoints TO PUBLIC;
 create or replace view sys.index_columns
 as
 select i.indrelid::integer as object_id
-  , CAST(CASE WHEN i.indisclustered THEN 1 ELSE 1+row_number() OVER(PARTITION BY c.oid) END AS INTEGER) AS index_id
+  , imap.index_id
   , a.attrelid::integer as index_column_id
-  , a.attnum::integer as column_id
+  , col.attnum::integer as column_id
   , a.attnum::sys.tinyint as key_ordinal
   , 0::sys.tinyint as partition_ordinal
   , 0::sys.bit as is_descending_key
   , 1::sys.bit as is_included_column
 from pg_index as i
 inner join pg_catalog.pg_attribute a on i.indexrelid = a.attrelid
+inner join pg_catalog.pg_attribute col on col.attname = a.attname and col.attrelid = i.indrelid
 inner join pg_class c on i.indrelid = c.oid
 inner join sys.schemas sch on sch.schema_id = c.relnamespace
+inner join (select indexrelid, CAST(CASE WHEN indisclustered THEN 1 ELSE 1+row_number() OVER(PARTITION BY indrelid) END AS INTEGER) AS index_id from pg_index) as imap on imap.indexrelid = i.indexrelid
```

Comment:
Can this be optimised by rather doing a join on pg_index and then leaving `CAST(CASE WHEN indisclustered THEN 1 ELSE 1+row_number() OVER(PARTITION BY indrelid) END AS INTEGER)`  as it was before, part of SELECT index_id.
The above mentioned would be a better optimisation _BUT_ if not possible we can think of adding this as a materialized CTE instead.
The performance of these catalog views gets exacerbated when we join multiple tsql views in a client facing query(with a lot of objects). These queries can also start taking as much time in the order of minutes and even hours. Hence, we should be very careful while making such changes.

Reply:
On my side, I can't think of a alternative to get the correct row_number() value...

Reply:
I agree with Kushaal here. If we can use CTE then we should use it

Reply:
In the original definition
```
CAST(CASE 
WHEN i.indisclustered THEN 1 
ELSE 1+dense_rank() OVER(PARTITION BY i.indrelid ORDER BY i.indexrelid)
END AS INT) AS index_id,
```
Could this work ?
Rank gives unique values to rows with same oder by column value, and dense rank makes sure there are no gaps